src/core/lib/security/transport/client_auth_filter.cc

```diff
@@ -110,8 +108,51 @@ struct call_data {
   grpc_auth_metadata_context auth_md_context =
       grpc_auth_metadata_context();  // Zero-initialize the C struct.
   grpc_closure async_result_closure;
-  grpc_closure check_call_host_cancel_closure;
-  grpc_closure get_request_metadata_cancel_closure;
+
+  // Mutex guarding async cancellation.
+  //
+  // There are two async calls done in this filter: check_call_host()
+  // and get_request_metadata(), both invoked while holding the call
+  // combiner.  The client_auth_pre_cancel_call() function, which is
+  // *not* run in the call combiner, needs to cancel either of these
+  // operations if they are in flight, so that we can quickly yield the
+  // call combiner and thus allow the cancel_stream op to come down in a
+  // timely manner.  In both cases, if the async operation has already
+  // completed, then a cancellation of that operation is a no-op.
+  //
+  // If client_auth_pre_cancel_call() runs before check_call_host() is
+  // started, it will set pre_cancelled.  When the code in the call
+  // combiner sees that pre_cancelled is set, it will yield the call
+  // combiner instead of invoking check_call_host().
+  //
+  // If client_auth_pre_cancel_call() runs after check_call_host() is
+  // started but before get_request_metadata() is started, it will
+  // attempt to cancel check_call_host() (which may be a no-op) and
+  // set pre_cancelled.  When the code in the call combiner sees that
+  // pre_cancelled is set, it will yield the call combiner instead of
+  // invoking get_request_metadata().
+  //
+  // If client_auth_pre_cancel_call() runs after get_request_metadata()
+  // is started, it will attempt to cancel get_request_metadata() (which
+  // may be a no-op).
+  //
+  // This mutex should not cause contention *except* when a cancellation
+  // is occurring, because all accesses to it other than in
+  // client_auth_pre_cancel_call() are done while holding the call combiner.
+  grpc_core::Mutex pre_cancel_mu;
+  bool pre_cancelled ABSL_GUARDED_BY(pre_cancel_mu) = false;
+  bool check_call_host_started ABSL_GUARDED_BY(pre_cancel_mu) = false;
+  bool get_request_metadata_started ABSL_GUARDED_BY(pre_cancel_mu) = false;
+
+  // The error seen from a cancel_stream op.
+  grpc_error_handle cancel_error = GRPC_ERROR_NONE;
+
+  // The batch containing send_initial_metadata is stored here if we
+  // receive a pre-cancellation but have not yet seen the cancel_stream
+  // op.  This ensures that the batch is failed with the right error,
+  // which may affect the call's status if the batch also contains
+  // recv_trailing_metadata.
+  grpc_transport_stream_op_batch* send_initial_metadata_batch = nullptr;
```

Comment:
why not just save the error instead and cancel the batch immediately?

Reply:
Because the error passed to `pre_cancel_call()` might not be the same as the error sent down with the `cancel_stream` op, and it's the latter value that we need to use for the `recv_trailing_metadata` batch.  (I don't remember at this point which test(s) it was that surfaced this problem, but there were test failures that required this fix.)

Reply:
Why would it not be the same? Isn't the trigger for this in call.cc? Don't we use the same error for `pre_cancel_call` and `cancel_stream`?

Reply:
You're right, we do.  Sorry, I must have been thinking of an earlier variant of this code when I wrote that.

I looked back at which test failure prompted this change to jog my memory, and it was this one:

https://source.cloud.google.com/results/invocations/ed092efc-40e8-496d-ae9f-013e2021a8cb/targets/%2F%2Ftest%2Fcpp%2Fend2end:end2end_test@poller%3Depoll1/tests;group=ProxyEnd2end%2FProxyEnd2endTest;test=RpcDeadlineExpires%2F0;row=1

The test was seeing UNKNOWN instead of DEADLINE_EXCEEDED.  I think the problem was that the batch containing `recv_trailing_metadata` had not yet hit the filter at the point at which `pre_cancel_call()` was invoked (because `pre_cancel_call()` is not invoked from within the call combiner, so the two are not synchronized from the perspective of the filter stack).  So that batch wound up getting failed with a different error, and that's the error that dictates the status code of the call.

IIRC, the actual problem that caused that test failure was in the client_channel filter, which I fixed in 91c43c8, but then I realized that the client_auth filter would have the same problem, so I fixed it (and some other edge cases I noticed) in 7b50bdc.  It might be that none of our tests would actually trigger this case in the client_auth filter, but it still seems safer to handle it properly.

Reply:
I don't think saving the previous batch is the right approach here. We should instead check who is responsible for invoking `recv_trailing_metadata` with the correct status/error and make sure that that error is being propagated.